### PR TITLE
teams: init at 1.2.00.32451

### PR DESCRIPTION
**Pull request by liff**
###### Motivation for this change

A Linux version of Microsoft Teams [was released](https://aka.ms/get-teams-linux).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notes

It’s somewhat similar to Slack so I was using the slack derivation as a reference.

### CODE REVIEW

**Comment by Elyhaka** [comment]
When trying to execute this package, It crashes with the following stacktrace :

```
> teams
A JavaScript error occurred in the main process
Uncaught Exception:
Error: libstdc++.so.6: cannot open shared object file: No such file or directory
    at process.module.(anonymous function) [as dlopen] (ELECTRON_ASAR.js:143:31)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:722:18)
    at Object.module.(anonymous function) [as .node] (ELECTRON_ASAR.js:152:18)
    at Module.load (internal/modules/cjs/loader.js:602:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:541:12)
    at Function.Module._load (internal/modules/cjs/loader.js:533:3)
    at Module.require (internal/modules/cjs/loader.js:640:17)
    at require (/nix/store/w6w2r130ma4aqcn5inxgnvqcjcv5rqzj-teams-1.2.00.32451/opt/teams/resources/app.asar/external/v8-compile-cache/v8-compile-cache.js:173:28)
    at Object.<anonymous> (/nix/store/w6w2r130ma4aqcn5inxgnvqcjcv5rqzj-teams-1.2.00.32451/opt/teams/resources/app.asar/node_modules/native-utils/index.js:1:173)
    at Object.<anonymous> (/nix/store/w6w2r130ma4aqcn5inxgnvqcjcv5rqzj-teams-1.2.00.32451/opt/teams/resources/app.asar/node_modules/native-utils/index.js:3:3)
```

The [slack derivation seems to address this issue at line 50](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/instant-messengers/slack/default.nix#L50). I will try to do it locally when I get back from work.

File: pkgs/applications/networking/instant-messengers/teams/default.nix
Comment (Elyhaka) [inline pkgs/applications/networking/instant-messengers/teams/default.nix]: I have been able to start Teams by adding this to the `modulePath`. I do not know if it is a good practice (Slack derivation seems to use it).

```suggestion
  ] + ":${stdenv.cc.cc.lib}/lib64";
```

Comment (liff) [reply pkgs/applications/networking/instant-messengers/teams/default.nix]: Thanks, I added it to `rpath` like in the Slack derivation rather than `LD_LIBRARY_PATH`. Does that work equally well for you? For some reason I cannot reproduce the error.

**Comment by liff** [comment]
Thanks for testing, @Elyhaka. Wonder if I should close this in favor of #75525 though?

**Comment by worldofpeace** [comment]
They're larely similar, but https://github.com/NixOS/nixpkgs/pull/75525/files uses autoPatchelfHook and wrapGAppsHook which i would've suggested here as well. Though the other is a draft.

File: pkgs/applications/networking/instant-messengers/teams/default.nix
Comment (Elyhaka) [reply pkgs/applications/networking/instant-messengers/teams/default.nix]: Strangely, it does not work with the addition to `rpath`, but it does work with it added to `modulePath`. I'm not really sure why, I'm curious to understand why it behaves this way. I'll do some `strace` digging.

**Comment by liff** [comment]
Changed to use `wrapGAppsHook` and `autoPatchelfHook`, and added some dependencies caught by the latter.

**Comment by Elyhaka** [comment]
Great ! It works perfectly for me with this update (no more `libstdc++` issue) :)

File: pkgs/applications/networking/instant-messengers/teams/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/instant-messengers/teams/default.nix]: this assert is completely unneeded, there's meta.platforms

File: pkgs/applications/networking/instant-messengers/teams/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/instant-messengers/teams/default.nix]:
```suggestion
stdenv.mkDerivation rec {
```

this allows self-referencing in the set, so you can move `version` into it and not inherit it.

File: pkgs/applications/networking/instant-messengers/teams/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/instant-messengers/teams/default.nix]: can we just reuse `atomEnv.packages`? We probably need to move that into a better attribute.

**Comment by worldofpeace** [comment]
This expression looks pretty ready @liff :+1:

File: pkgs/applications/networking/instant-messengers/teams/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/instant-messengers/teams/default.nix]: can we add downloadPage = "https://teams.microsoft.com/downloads";?

**Comment by liff** [comment]
Since #75525 was closed, should I add @jonringer as a maintainer?

**Review by worldofpeace — APPROVED** [review]
LGTM. I will defer to people who actually use the application.

**Review by aanderse — APPROVED** [review]
I have tested this PR and the software works as expected. Thank you very much @liff :tada: 

@jonringer I'll leave it to you to convince whomever you need to convince that we need a linux port of outlook next :wink: :laughing:

**Comment by jonringer** [comment]
not sure why, but i was having problems on my workstation, home computer it works fine though

**Comment by jonringer** [comment]
I just did a call using this, the audio is very poor, not entire sure why. maybe using libpulseaudio will help remedy that

**Comment by jonringer** [comment]
also, it was eating my close application, :(

**Comment by aanderse** [comment]
@jonringer did not use for phone call but used during the day for regular stuff and so far no issues... What is close application?

**Comment by jonringer** [comment]
If I try to click the X in the upper right-hand corner

**Comment by aanderse** [comment]
Oh sorry I understand... I was able to click and then I received notification that `teams` wouldn't exit but remain in tray, as expected.

Maybe I'll try to install on another device :thinking:

**Comment by jonringer** [comment]
The other thing is that, I'm only using lightdm+i3, and no desktop environment. Probably the cause of the wonky issues.

**Comment by aanderse** [comment]
Yeah definitely might be. I'm on `plasma`.

I'll test voice call tomorrow.

**Pull request merged**
